File: gin/library/safety.go
package library

import (
	"regexp"
	"webmis/config"

	"github.com/dgrijalva/jwt-go"
)

/* 验证类 */
type Safety struct{}

/* 正则-公共 */
func (s Safety) IsRight(name string, val string) bool {
	data := map[string]string{
		"uname":  "^[a-zA-Z][a-zA-Z0-9\\_\\@\\-\\*\\&]{3,15}$",
		"passwd": "^[a-zA-Z0-9|_|@|-|*|&]{6,16}$",
		"tel":    "^1\\d{10}$",
		"email":  "^([a-zA-Z0-9_-])+@([a-zA-Z0-9_-])+(\\.[a-zA-Z0-9_-])+",
		"idcard": "^[1-9]\\d{7}((0\\d)|(1[0-2]))(([0|1|2]\\d)|3[0-1])\\d{3}$|^[1-9]\\d{5}[1-9]\\d{3}((0\\d)|(1[0-2]))(([0|1|2]\\d)|3[0-1])\\d{3}([0-9]|X)$",
	}
	res, _ := regexp.MatchString(data[name], val)
	return res
}

/* 正则-验证 */
func (s Safety) Test(reg string, val string) bool {
	res, _ := regexp.MatchString(reg, val)
	return res
}

/* 加密-JWT */
func (s Safety) Encode(param map[string]interface{}) (string, error) {
	cfg := config.Env() //配置
	token := jwt.New(jwt.SigningMethodHS256)
	claims := make(jwt.MapClaims)
	for index, val := range param {
		claims[index] = val
	}
	token.Claims = claims
	res, err := token.SignedString([]byte(cfg.Key))
	return res, err
}

/* 解密-JWT */
func (s Safety) Decode(token string) (jwt.MapClaims, error) {
	// 配置
	cfg := config.Env()
	// 解析
	res, err := jwt.Parse(token, func(t *jwt.Token) (interface{}, error) {
		if _, ok := t.Method.(*jwt.SigningMethodHMAC); !ok {
			return nil, nil
		}
		return []byte(cfg.Key), nil
	})
	if err != nil {
		return nil, err
	}
	// 结果
	if claims, ok := res.Claims.(jwt.MapClaims); ok && res.Valid {
		return claims, err
	}
	return nil, nil
}

File: java/src/main/java/webmis/library/Safety.java
package webmis.library;

import java.util.HashMap;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import webmis.config.Env;

import com.alibaba.fastjson.JSON;
import com.alibaba.fastjson.TypeReference;
import com.auth0.jwt.JWT;
import com.auth0.jwt.algorithms.Algorithm;
import com.auth0.jwt.interfaces.DecodedJWT;
import com.auth0.jwt.exceptions.JWTCreationException;
import com.auth0.jwt.exceptions.JWTDecodeException;

/* 验证类 */
public class Safety {

  /* 正则-公共 */
  public static boolean IsRight(String name, String val){
    HashMap<String,Object> data = new HashMap<String,Object>();
    data.put("uname", "^[a-zA-Z][a-zA-Z0-9\\_\\@\\-\\*\\&]{3,15}$");
    data.put("passwd", "^[a-zA-Z0-9|_|@|-|*|&]{6,16}$");
    data.put("tel", "^1\\d{10}$");
    data.put("email", "^([a-zA-Z0-9_-])+@([a-zA-Z0-9_-])+(\\.[a-zA-Z0-9_-])+");
    data.put("idcard", "^[1-9]\\d{7}((0\\d)|(1[0-2]))(([0|1|2]\\d)|3[0-1])\\d{3}$|^[1-9]\\d{5}[1-9]\\d{3}((0\\d)|(1[0-2]))(([0|1|2]\\d)|3[0-1])\\d{3}([0-9]|X)$");
    Pattern pattern = Pattern.compile(data.get(name).toString());
    Matcher matcher = pattern.matcher(val);
    return matcher.matches();
  }

  /* 正则-验证 */
  public static boolean Test(String reg, String val){
    Pattern pattern = Pattern.compile(reg);
    Matcher matcher = pattern.matcher(val);
    return matcher.matches();
  }

  /* 加密 */
  public static String Encode(HashMap<String,Object> data){
    String token = null;
    try {
      Algorithm algorithm = Algorithm.HMAC256(Env.key);
      token = JWT.create().withIssuer(JSON.toJSONString(data)).sign(algorithm);
      return token;
    } catch (JWTCreationException exception){
      return token;
    }
  }

  /* 解密 */
  public static HashMap<String,Object> Decode(String token){
    HashMap<String,Object> res = null;
    try {
      DecodedJWT jwt = JWT.decode(token);
      String data = jwt.getIssuer();
      return JSON.parseObject(data,new TypeReference<HashMap<String,Object>>(){});
    } catch (JWTDecodeException exception){
      return res;
    }
  }
  
}
